### PR TITLE
feat(backend): provision private key

### LINKED ISSUE #606: Self-provisioned keys
When a Rafiki instance calls out to other open payment endpoints, it needs to provide a key for getting a grant. This is required to identify itself. Hence, the Rafiki instance needs to provision its own key(s).

### PR DESCRIPTION
**Pull request by sabineschaller**
## Changes proposed in this pull request

- load private key or create one
- part of backend `config`

## Context

- fixes #606 

## Checklist

- [x] Related issues linked using `fixes #number`
- [ ] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass

### CODE REVIEW

**Review by wilsonianb — DISMISSED** [review]

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:156]: Actually...
should this be storing the generated key to at least a tmp file?

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:145]: Should this include support for reading an encrypted key file?

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:156]: I was wondering something similar and wanted to make a comment but then had to run. 
I was wondering whether we should store the keys somewhere when Rafiki "goes down" for whatever reason. Is that what you want to use the tmp file for?
Also, where would you like to put the tmp file?

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:145]: Yes, probably. I'll include that

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.ts:156]: I don't have a location in mind, but maybe whatever we choose could be the default value for `PRIVATE_KEY_PATH`

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:145]: Now that I thought more about it, I'm not so sure. We would also have to specify which encryption mechanism to use. And those key files are usually not encrypted. 🤔

Comment (wilsonianb) [reply packages/backend/src/config/app.ts:145]: I'm fine leaving it out for now.

Comment (sabineschaller) [reply packages/backend/src/config/app.ts]: #642

**Comment by sabineschaller** [comment]
Should this PR also add the key to the internal key registry?

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:157]: This fails if the directory already exists

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:113]:
```suggestion
    envString('PRIVATE_KEY_PATH', './tmp/private.pem')
```
If the `backend` is run from a working directory like `/root` the target directory isn't possible.

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:159]: Maybe make a const string used both here and in the `privateKey` assignment above

Comment (sabineschaller) [reply packages/backend/src/config/app.ts]: Which `privateKey` assignment?

Comment (wilsonianb) [reply packages/backend/src/config/app.ts]: https://github.com/interledger/rafiki/pull/638/files#diff-6300c4703a2fb930745656aed4f5d68507eb07a20ae1f817a1ca3c9d90901facR112

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:144]: Is it worth adding a `config.test.ts` for this?

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.ts]: maybe something like
```
const TMP_DIR = './tmp'
const PRIVATE_KEY_FILE = `${TMP_DIR}/private-key.pem`

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:160]: I was wondering that, too. But I don't see a reason to keep the old ones as long as the new one is also associated with this RS. And by associating I mean adding it to the internal key registry. Should this PR also do that?

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:144]: Probably yes 👍

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts]: Oh, that far above 😄

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:160]: Well, we'll need the key itself for the signature, won't we? But I agree, we should also provide the jwk.

File: packages/backend/src/config/app.test.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.test.ts:24]:
```suggestion
```suggestion
      expect(key).toMatchObject({
        asymmetricKeyType: 'ed25519',
        type: 'private'
      })
      expect(jwk).toEqual(key.export({ format: 'jwk' }))
      expect(jwk).toEqual({
```

File: packages/backend/src/config/app.test.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.test.ts:52]: https://nodejs.org/api/crypto.html#keyobjectequalsotherkeyobject
```suggestion
      assert.ok(key.equals(keypair.privateKey))
      expect(jwk).toEqual(key.export({ format: 'jwk' }))
      expect(jwk).toEqual({
        crv: 'Ed25519',
        kty: 'OKP',
        d: expect.any(String),
        x: expect.any(String)
      })
```

File: packages/backend/src/config/app.test.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.test.ts:30]:
```suggestion
      expect(fs.readFileSync(PRIVATE_KEY_FILE, 'utf8')).toEqual(
        key.export({ format: 'pem', type: 'pkcs8' })
      )
```

File: packages/backend/src/config/app.test.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.test.ts:20]: Test this with or without preexisting `TMP_DIR`

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:168]: Now I'm wondering if `parseOrProvisionKey` should only operate with `keyPath` as opposed to `TMP_DIR`/`PRIVATE_KEY_FILE`, to prevent an operator from not realizing that the default `PRIVATE_KEY_PATH` is being used and not their configured one (even though that is being logged).

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.ts:167]: I'm leaning towards not overwriting.
I don't like the idea of potentially overwriting a file not intended for for Rafiki, however unlikely such a misconfiguration may be.

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [inline packages/backend/src/config/app.ts:156]: Given that the `jwk` isn't actually being included in the config and I think`parseOrProvisionKey` doens't actually need to generate the `jwk`, I vote we drop it.
(I'm trying to see how we'd generate an ed25519 private `KeyObject` from just `randomBytes` below)

```suggestion
export function parseOrProvisionKey(keyPath: string): crypto.KeyObject {
  try {
    const key = crypto.createPrivateKey(fs.readFileSync(keyPath))
    if (key.asymmetricKeyType === 'ed25519') {
      return key
```

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.ts:167]: >And by associating I mean adding it to the internal key registry. Should this PR also do that?

I think we should hold off until we decide if we're going to switch the key registry to using JWKS
- https://github.com/interledger/open-payments/pull/190

File: packages/backend/src/config/app.test.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.test.ts:52]: `Property 'equals' does not exist on type 'KeyObject'.`

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:168]: I'm not sure I understand. `PRIVATE_KEY_FILE` does include the path.

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:167]: > I'm leaning towards not overwriting.
I don't like the idea of potentially overwriting a file not intended for for Rafiki, however unlikely such a misconfiguration may be.

Should we use unique tmp dirs?

Comment (sabineschaller) [reply packages/backend/src/config/app.ts:167]: > I think we should hold off until we decide if we're going to switch the key registry to using JWKS

I'll create an issue to also create a payment pointer for the Rafiki instance itself that will then be linked to the key, in whatever way and form we end up doing that (probably JWKS endpoint).

File: packages/backend/.gitignore
Comment (wilsonianb) [reply packages/backend/.gitignore:6]: *re-add`

File: packages/backend/src/config/app.test.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.test.ts]: :confused: 
https://nodejs.org/dist/latest-v16.x/docs/api/crypto.html#keyobjectequalsotherkeyobject
Is that a typescript build error?

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.ts:168]: As in `parseOrProvisionKey` would only read or write (maybe not overwrite) the file at `keyPath`.
As is, it reads from `keyPath` but writes to `PRIVATE_KEY_FILE`.

File: packages/backend/src/config/app.test.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.test.ts]: yes
```
src/config/app.test.ts:59:21 - error TS2339: Property 'equals' does not exist on type 'KeyObject'.
59       assert.ok(key.equals(keypair.privateKey))
                       ~~~~~~
Found 1 error.
```
I even made sure to use node v16.18.0.

File: packages/backend/src/config/app.ts
Comment (sabineschaller) [reply packages/backend/src/config/app.ts:165]: But what if it only has read access and not write access at `keyPath`?

Comment (wilsonianb) [reply packages/backend/src/config/app.ts:165]: I'm imagining an operating running multiple instances of rafiki `backend` in a k8s cluster and configuring `PRIVATE_KEY_FILE` as a path to a secret.
In which case, I think we should throw if we can't use (or write to) the configured file (`keyPath`).
Otherwise you end up with individual instances provisioning their own private keys.

**Review by mkurapov — APPROVED** [review]
I was able to test successfully, both with existing and new, generated key. 

As you two are discussing in the comment, similarly, I do see the benefit of maybe having the `parseOrProvisionKey` take `undefined` as a possible argument, and doing the `keyPath` generation inside the method itself. This way if a defined `keyPath` _is_ given, we know its's coming from a user-set config, not a default value, in which case we will throw as @wilsonianb mentions

File: packages/backend/src/config/app.ts
Comment (mkurapov) [inline packages/backend/src/config/app.ts:153]: It would be nice to have `'Ed25519'` coming from a type declaration, but it looks like we are limited to just a string here.

**Review by mkurapov — APPROVED** [review]

File: packages/backend/src/config/app.ts
Comment (wilsonianb) [reply packages/backend/src/config/app.ts:153]: Maybe we could eventually do schema validation using:
https://github.com/interledger/open-payments/blob/b125873edb6d33acd40eb304e026e5111ec1dba9/openapi/shared/schemas.yaml#L201-L234

**Comment by wilsonianb** [comment]
Merging/rebasing main _should_ resolve the failing Docker checks

**Pull request merged**
